### PR TITLE
[SPARK-7008] An implement of Factorization Machines based on Scala and Spark MLlib.

### PR DESCRIPTION
**Pull request by zhengruifeng**
An implementation of Factorization Machines based on Scala and Spark MLlib.
Factorization Machine is a kind of state-of-the-art machine learning algorithm for multi-linear regression, and is widely used in recommendation systems.
Factorization Machines algorithm and its C++ implement LibFM works well in recent years' recommendation competitions.

A FMModel consist of three parts: 
    an Intercept (optional),
    an one-way interactions weights like other linear models (optional),
    a numFactors \* numFeatures matrix representing the factors of each feature (mandatory).

I implement the training algorithm with SGD provided by MLlib's GradientDescent (to use MLlib's SGD, the model is encoded to a dense vector during training).

Ref:
http://libfm.org/
http://doi.acm.org/10.1145/2168752.2168771
http://www.inf.uni-konstanz.de/~rendle/pdf/Rendle2010FM.pdf

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/regression/FactorizationMachine.scala [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+import org.apache.spark.rdd.RDD
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.mllib.util.Loader._
+import org.apache.spark.mllib.util.{Loader, Saveable}
+import org.apache.spark.sql.{DataFrame, SQLContext}
+
+/**
+ * Created by zrf on 4/13/15.
+ */
+
```
Comment (witgo) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/FactorizationMachine.scala]: Here should be deleted.

File: mllib/src/test/scala/org/apache/spark/mllib/regression/FactorizationMachineSuite.scala [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+import org.scalatest.FunSuite
+import scala.util.Random
+
+import org.apache.spark.mllib.linalg.{Matrix, DenseMatrix, Vectors}
+import org.apache.spark.mllib.util.{LocalClusterSparkContext, MLlibTestSparkContext}
+import org.apache.spark.util.Utils
+
+/**
+ * Created by zrf on 4/20/15.
+ */
```
Comment (witgo) [inline mllib/src/test/scala/org/apache/spark/mllib/regression/FactorizationMachineSuite.scala]: It should be deleted.

**Comment by mengxr** [comment]
@zhengruifeng Thanks for submitting your impl as a Spark package at http://spark-packages.org/package/zhengruifeng/spark-libFM! As we discussed on the JIRA, we need some time to evaluate FM and discuss a scalable solution. So do you mind closing this PR for now?

**Comment by zhengruifeng** [comment]
OK, please close it.

**Comment by withDataman** [comment]
hello, why don't load FM model?

**Pull request closed without merging**